module/zfs/qat_compress.c

```diff
@@ -464,20 +489,27 @@ qat_compress(qat_compress_dir_t dir, char *src, int src_len,
 	}
 
 	if (out_pages) {
-		for (page_num = 0;
-		    page_num < buf_list_dst->numBuffers;
-		    page_num++) {
+		for (page_num = 0; page_num < dst_pages; page_num++) {
 			kunmap(out_pages[page_num]);
 		}
 		QAT_PHYS_CONTIG_FREE(out_pages);
 	}
 
+	if (add_pages) {
+		for (page_num = 0;
+		    page_num < buf_list_dst->numBuffers - dst_pages;
+		    page_num++) {
+			kunmap(add_pages[page_num]);
+		}
+		QAT_PHYS_CONTIG_FREE(add_pages);
+	}
+
 	QAT_PHYS_CONTIG_FREE(buffer_meta_src);
 	QAT_PHYS_CONTIG_FREE(buffer_meta_dst);
 	QAT_PHYS_CONTIG_FREE(buf_list_src);
 	QAT_PHYS_CONTIG_FREE(buf_list_dst);
 
-	return (ret);
+	return (status);
```

Comment:
"status" initial value is success:
CpaStatus status = CPA_STATUS_SUCCESS
some fail branch will return success.

Reply:
The idea here was that CPA_SATUS_FAIL would indicate a real hardware failure whereas a `src_len == dst_len` would indicate that compression simply wasn't worth it

Reply:
Previously the function could return `-1` but the calling code would check that it was equal to `CPA_STATUS_SUCCESS`

Reply:
OK - I see, but for some fail cases, it will return SUCCESS?

Reply:
I think it makes sense. Perhaps @behlendorf has a better idea of how we should convey this meaning?